### PR TITLE
implementing operations

### PR DESCRIPTION
**Pull request by benjaglobant**
Implementing operations (plus, substraction, multiply and division). Showing error messages when the user trys to:
        Divide by zero.
        Put more than one point in the same number.
        Put two or more consecutive operators.
        Put an operator without number.
 Also i made the test for every functionalities of the calculator. Testing the model from the presenter and the view mocked using mockito-kotlin.
Gif showing the error messages:
calcuKotlinFinish
Images of the test and its coverage:
image
image

### CODE REVIEW

File: app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt [hunk 1]
```diff
@@ -62,17 +73,60 @@ class CalculatorPresenter(
...
+                    view.showMessage(CalculatorError.MATH_ERROR)
+                    onClearButtonPressed()
+                }
+            }
+        }
+        return EMPTY_STRING
+    }
+
+    override fun onEqualPressed() {
+        if (!((model.first_operand.isEmpty()) || (model.second_operand.isEmpty()) || (model.operator == EMPTY_CHAR))) {
```
Comment (SartiMau) [inline app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt]: you can use here 
((model.first_operand.isNotEmpty()) && (model.second_operand.isNotEmpty())
instad of
!((model.first_operand.isEmpty()) || (model.second_operand.isEmpty())

if I'm not wrong it gives more performance

Comment (benjaglobant) [reply app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt]: Yes, you are right. Thanks Mau, done!

File: app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt [hunk 2]
```diff
@@ -62,17 +73,60 @@ class CalculatorPresenter(
...
+            OPERATOR_DIVIDE -> {
+                if (!model.second_operand.equals(NUMBER_ZERO)) {
+                    return decimalFormat.format(model.first_operand.toDouble() / model.second_operand.toDouble())
+                } else {
+                    view.showMessage(CalculatorError.MATH_ERROR)
+                    onClearButtonPressed()
+                }
+            }
+        }
+        return EMPTY_STRING
```
Comment (nicolbidone) [inline app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt]: here you can use https://kotlinlang.org/docs/reference/control-flow.html#when-expression

Comment (benjaglobant) [reply app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt]: Done Nico, thanks for your help!

File: app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt [hunk 3]
```diff
@@ -1,16 +1,25 @@
...
+import com.globant.onboardingcalculatorwithkotlin.utils.Constants.OPERATOR_PLUS
+import com.globant.onboardingcalculatorwithkotlin.utils.Constants.OPERATOR_SUBSTRACTION
+import java.text.DecimalFormat
 
 class CalculatorPresenter(
     private val model: CalculatorContracts.Model,
     private val view: CalculatorContracts.View
 ) : CalculatorContracts.Presenter {
 
+    //private val decimalFormat: DecimalFormat = DecimalFormat(DECIMAL_FORMAT)
```
Comment (SartiMau) [inline app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt]: if this is not necessary, please remove it

File: app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt [hunk 4]
```diff
@@ -62,17 +73,63 @@ class CalculatorPresenter(
...
             } else {
-                view.showDecimalError()
+                view.showMessage(CalculatorError.DECIMAL_ERROR)
             }
         }
         updateVisor()
     }
+
+    private fun validOperation(): Boolean {
+        if (!model.second_operand.equals(NUMBER_ZERO))
```
Comment (SartiMau) [inline app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt]: you can put here 

return !model.second_operand.equals(NUMBER_ZERO)

File: app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt [hunk 5]
```diff
@@ -62,17 +73,63 @@ class CalculatorPresenter(
...
+            }else{
+                view.clearVisor()
+            }
+        } else {
+            view.showMessage(CalculatorError.INCOMPLETE_OPERATION)
+        }
+    }
+}
+
+enum class CalculatorError(val message: Int) {
```
Comment (mauroLonghi) [inline app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt]: this value (val message: Int) its neccesary?

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/test/java/com/globant/onboardingcalculatorwithkotlin/CalculatorPresenterTest.kt [hunk 6]
```diff
@@ -23,6 +23,7 @@ import com.nhaarman.mockitokotlin2.verify
 import org.junit.Test
 import org.junit.Assert.assertEquals
 
+const val WRONG_OPERATOR: Char = '?'
```
Comment (nicolbidone) [inline app/src/test/java/com/globant/onboardingcalculatorwithkotlin/CalculatorPresenterTest.kt]: if this is not used outside the test, it should be inside the companion object

File: app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt (hunk 5 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt]: Refactored!

File: app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt (hunk 3 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt]: Done!!

File: app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt (hunk 4 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt]: Done!

File: app/src/test/java/com/globant/onboardingcalculatorwithkotlin/CalculatorPresenterTest.kt (hunk 6 above)
Comment (benjaglobant) [reply app/src/test/java/com/globant/onboardingcalculatorwithkotlin/CalculatorPresenterTest.kt]: Done!

**Review by mauroLonghi — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]

**Review by nicolbidone — APPROVED** [review]

**Pull request merged**